---
title: "palyginimas tarp normalizavimo metodu, gtex"
author: "Ieva"
date: "2023-08-24"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(glmnet)
library("TCGAbiolinks")
library(DESeq2)
library(biomaRt)
library(org.Hs.eg.db)
library(clusterProfiler)
library(WGCNA)
library(gridExtra)
library(CorLevelPlot)
library(venn)
library(GDCRNATools)
```

# Setup
pradedu nuo step01 sujungtu duomenu
# Count data and normalize
pirmiausia reikia count matricos su rows as genes
```{r}
full_train <- readRDS("C:/Users/Ieva/Desktop/NVI GDL/R projetcs/tcga-ov-data/TCGA-OV-mRNA data/mRNA and clinical data TCGA-OV/train_gtcga.RDS")
# regain count data: genes in rows
counts_gtcga <- full_train[, 48:35165] #nusimažinu tik countus, bet dar liko gtex
which( colnames(counts_gtcga)=="gtex" ) #35118 stulpelis
counts_gtcga <- counts_gtcga[, -35118]
#transpose!
counts_gtcga <- t(counts_gtcga) #large numeric matrix with rows as genes
```

coldata/phenodata, rows as samples
```{r}
which(colnames(full_train)=="gtex" ) #35118 stulpelis
coldata_gtcga <- full_train[, c(1:47, 35165)]
colnames(coldata_gtcga) #48
```

outlier deletion (from the step01)
```{r}
### NOTE: If there are batch effects observed, correct for them before moving ahead

# exclude outlier samples
samples.to.be.excluded <- c('TCGA-61-2094-01A-01R-1568-13', 
                          "TCGA-13-0760-01A-01R-1564-13", 
                          "TCGA-24-1428-01A-01R-1564-13",
                          "TCGA-24-1845-01A-01R-1567-13",
                          "TCGA-13-1482-01A-01R-1565-13",
                          "GTEX-WXYG-1426-SM-4ONCK")
counts_gtcga <- counts_gtcga[,!(colnames(counts_gtcga) %in% samples.to.be.excluded)] #lieka 488 zmones bendrai

# exclude outlier samples
coldata_gtcga <- coldata_gtcga %>% 
  filter(!row.names(.) %in% samples.to.be.excluded) #lieka 488 zmones

all(rownames(coldata_gtcga) %in% colnames(counts_gtcga))
nrow(coldata_gtcga) == ncol(counts_gtcga)
all(rownames(coldata_gtcga) == colnames(counts_gtcga)) #reikia reagangint
```
# normalize GDCRNAtools: gdcVoomNormalization
šito metodo esmė yra ta kad daro edgeR metodą - TMM trimmed mean of values,
o tada su limma voom transformacija, logoritmuoja, finina curve ir iš jų gauna "svorius" su kuriais normalizuoja, todėl šitas metodas daugiausiai dalykų pakeičia.
In the end gauni skaičius, kurie akivaizdžiai logoritmuoti, decimals, teigiami/neigiami
```{r}
#1 type 2: from GDCRNAtools: gdcVoomNormalization: does TNB ir voom transformatio
trainmRNA_voom <- gdcVoomNormalization(counts = counts_gtcga, filter = FALSE) #60660 genes
#pridedam normal vardus
trainmRNA_voomt <- t(trainmRNA_voom)
```
now for the glm su limma vadinu bet ištiesu GDCrna_tools
## lasoo?
```{r}
dim(trainmRNA_voom) #buvo tiek: 35117   488
train_response <- as.factor(coldata_gtcga$gtex)
res_gtex = cv.glmnet(
  x = trainmRNA_voomt,
  y = train_response,
  alpha = 1,
  family = "binomial"
)
res_gtex #12 visur
```
## elastic net?
```{r}
dim(trainmRNA_voom) #buvo tiek: 35117   488
train_response <- as.factor(coldata_gtcga$gtex)
res_gtex_0.5 = cv.glmnet(
  x = trainmRNA_voomt,
  y = train_response,
  alpha = 0.5,
  family = "binomial"
)
res_gtex_0.5 #207 omg
```
```{r}
# Getting genes that contribute for the prediction
res_coef_gtex = coef(res_gtex, s="lambda.min") # the "coef" function returns a sparse matrix
res_coef_gtex = res_coef_gtex[res_coef_gtex[,1] != 0,] 
res_coef_gtex = res_coef_gtex[-1]
res_coef_gtex_names = names(res_coef_gtex) # get names of the (non-zero) variables.

```
## biomart
```{r}
# cia uztenka karta padaryt
listEnsembl() #shows available databases
ensembl <- useEnsembl(biomart = "genes") #unsuportive gali but
datasets <- listDatasets(ensembl) #issirinkti human genes
ensembl.con <- useMart("ensembl", dataset = 'hsapiens_gene_ensembl') #name of data base ir data set
attr <- listAttributes(ensembl.con)
filters <- listFilters(ensembl.con)
```
## gdc genes?
```{r}
#convert relavant gene names
res_coef_gtex_names <- gsub("\\..*", "",res_coef_gtex_names)
res_coef_gtex_names
#biomart
gtcga_genes <- getBM(attributes = c('ensembl_gene_id','external_gene_name'), #values to retreve
      filters = "ensembl_gene_id", #input on quary
      values = res_coef_gtex_names,
      mart = ensembl.con) #sukurtas conection object
gtcga_genes
```
okay nematyti negirdėti genai let's go
```{r}
#from org.Hs.eg.db
#need to remove na, jei ju yr, dabar ner
gtcga_genes.symbols <- gtcga_genes$external_gene_name
gtcga_genes.entrez <- mapIds(org.Hs.eg.db, gtcga_genes.symbols, 'ENTREZID', 'SYMBOL')
#names chr vector of entrez ids
ggo.gtcga <- groupGO(gene = gtcga_genes.entrez,
               OrgDb    = org.Hs.eg.db,
               ont      = "CC",
               level    = 3,
               readable = TRUE)

ggo.gtcga <- as.data.frame(ggo.gtcga) # count is 0 because i provided no counts!
ggo.gtcga_filt <- ggo.gtcga %>% filter(Count != "0")
ggo.gtcga_filt
```
```{r}
#ego = enrichment go
ego.gtcga <- enrichGO(gene = gtcga_genes.entrez,
                OrgDb         = org.Hs.eg.db,
                ont           = "CC",
                pAdjustMethod = "BH",
                pvalueCutoff  = 0.05,
                qvalueCutoff  = 0.10,
        readable      = TRUE)
as.data.frame(ego.gtcga) #nothing selected
```
## final gdc list and functions
i guess back to manual from this:

1  ENSG00000142694              EVA1B   Eva-1 homolog B, maybe immune function
2  ENSG00000182196            ARL6IP4   ADP Ribosylation Factor Like GTPase 6 Interacting Protein 4, maybe a splicing regulator
3  ENSG00000184205             TSPYL2   G1 chekpoint
4  ENSG00000207165            SNORA70   RNU70, small nucleolar RNA, H/ACA box 70
5  ENSG00000210196              MT-TP   mitochondrial transfer RNA (tRNA) proline
6  ENSG00000215837             SDHAP2   ccinate dehydrogenase complex flavoprotein subunit A pseudogene
7  ENSG00000226232           NPIPB14P   Nuclear Pore Complex Interacting Protein Family Member B14 Pseudogene.
8  ENSG00000231503             PTMAP4   pseudogene
9  ENSG00000253797             UTP14C   a ribosome processome component
10 ENSG00000265681              RPL17   ribosome 60s
11 ENSG00000266820            KPNA2P3   pseaudogene, unknown function
12 ENSG00000267368            UPK3BL1   Uroplakin -urothelium-specific transmembrane protein 

# TCGAanalyze_Normalization
šio metodo esmė yra EDAseq package pritaikyt ir normalizuot genų ilgiui/ gc skaičiam ir panašiem dalykam.
lieka sveiki skaičiai ir iš lyginimo su raw counts no idea ar ka padaro, bet next step nusifiltruot biškuti duomenu tai at least 0 nelieka 
```{r}
trainmRNA_tcga <- TCGAanalyze_Normalization(
  tabDF = counts_gtcga, 
  geneInfo =  geneInfoHT
) #(46462)
#clean low and high counts
# quantile filter of genes #34766 genes
trainmRNA_tcga_filtered <- TCGAanalyze_Filtering(
  tabDF = trainmRNA_tcga,
  method = "quantile", 
  qnt.cut =  0.25
) #(34834)

trainmRNA_tcga_filtered <- as.data.frame(trainmRNA_tcga_filtered)
trainmRNA_tcga_filtered <- t(trainmRNA_tcga_filtered)
```

```{r}
htree_norm <- hclust(dist(trainmRNA_tcga_filtered), method = "average") #can take some time
dend3 <- as.dendrogram(htree_norm)
col_aa_red <- ifelse(grepl("GTEX", labels(dend3)), "red", "blue")
dend4 <- assign_values_to_leaves_edgePar(dend=dend3, value = col_aa_red, edgePar = "col") 
pdf(file="C:/Users/Ieva/Desktop/NVI GDL/R projetcs/tcga-ov-data/TCGA-OV-mRNA data/Figures mRNA/htree_tcganal.pdf", height=80, width=100) #išsaugijimui didesniu formatu
plot(dend4) #dstant samples should be excluded #save portrait 20x66 inch pdf
dev.off()
```

## lasoo?
```{r}
dim(trainmRNA_tcga_filtered) #buvo tiek: 15864   488 (gerokai mažiau nei gdc metodu!)
train_response <- as.factor(coldata_gtcga$gtex)
res_gtex_tcga = cv.glmnet(
  x = trainmRNA_tcga_filtered,
  y = train_response,
  alpha = 1,
  family = "binomial"
)
res_gtex_tcga #25 visur
```
## elastic net?
```{r}
res_gtex_tcga_0.5 = cv.glmnet(
  x = trainmRNA_tcga_filtered,
  y = train_response,
  alpha = 0.5,
  family = "binomial"
)
res_gtex_tcga_0.5 #143 ir 130 for 1se 
```
## gdc genes?
```{r}
# Getting genes that contribute for the prediction
res_coef_gtex_tcga = coef(res_gtex_tcga, s="lambda.min") # the "coef" function returns a sparse matrix
res_coef_gtex_tcga = res_coef_gtex_tcga[res_coef_gtex_tcga[,1] != 0,] 
res_coef_gtex_tcga = res_coef_gtex_tcga[-1]
res_coef_gtex_tcga = names(res_coef_gtex_tcga) # get names of the (non-zero) variables.
```

```{r}
#convert relavant gene names
#no gsub in TCGA needed
res_coef_gtex_names
#biomart
gtcga_genes_tcga <- getBM(attributes = c('ensembl_gene_id','external_gene_name'), #values to retreve
      filters = "ensembl_gene_id", #input on quary
      values = res_coef_gtex_tcga,
      mart = ensembl.con) #sukurtas conection object
gtcga_genes_tcga
```
okayyy lots bet vardų
```{r}
gtcga_genes_tcga <- gtcga_genes_tcga[-c(9,10, 12, 16:25),]
tcga_vardai <- gtcga_genes_tcga$external_gene_name
gdc_vardai <- gtcga_genes$external_gene_name
intersect(tcga_vardai, gdc_vardai)
```
okay tai top 4 turim: 
5  ENSG00000142694              EVA1B Eva-1 homolog B, maybe immune function
6  ENSG00000182196            ARL6IP4 ADP Ribosylation Factor Like GTPase 6 Interacting Protein 4, maybe a splicing regulator
7  ENSG00000184205             TSPYL2  G1 chekpoint
15 ENSG00000253797             UTP14C  a ribosome processome component


kiti genai, nesutampantys tarp normalizavimu
  ensembl_gene_id external_gene_name
1  ENSG00000002822             MAD1L1   Mitotic Arrest Deficient 1 Like 1
2  ENSG00000076554              TPD52   Tumor Protein D52 ATM regulation
3  ENSG00000090539               CHRD   Chordin, vystymosi genas, su pusių nustatymu susijęs
4  ENSG00000102362              SYTL4   Synaptotagmin-like protein 4, molekuliu transporteris
8  ENSG00000198134             PTMAP9   Prothymosin Alpha Pseudogene 9.
11 ENSG00000230189          RABGEF1P2   RABGEF1 Pseudogene 2.
13 ENSG00000243410          PSMD6-AS1   SMD6 Antisense RNA 1, is an RNA gene that belongs to the long non-coding RNA (lncRNA) class
14 ENSG00000243749            TMEM35B   Transmembrane Protein 35B 

# dėl deseq, idk kaip ji pritaikyt nes iš karto turėtų modelį taikyt jis. 
su DESEQ, skaičiai decimals, bet ner minusinių
```{r}
# create dds
dds <- DESeqDataSetFromMatrix(countData = counts_gtcga,
                              colData = coldata_gtcga,
                              design = ~ 1) # not spcifying model

## remove all genes with counts < 15 in more than 75% of samples (488*0.75=366)
## suggested by WGCNA on RNAseq FAQ

dds75 <- dds[rowSums(counts(dds) >= 15) >= 366,] #čia nuo meginiu sk!
nrow(dds75) # 4684 genes


# perform variance stabilization
dds_norm <- vst(dds75) 


# get normalized counts
norm.counts <- assay(dds_norm) %>% 
  t()
```
```{r}
htree_norm2 <- hclust(dist(norm.counts), method = "average") #can take some time
dend3 <- as.dendrogram(htree_norm2)
col_aa_red <- ifelse(grepl("GTEX", labels(dend3)), "red", "blue")
dend5 <- assign_values_to_leaves_edgePar(dend=dend3, value = col_aa_red, edgePar = "col") 
pdf(file="C:/Users/Ieva/Desktop/NVI GDL/R projetcs/tcga-ov-data/TCGA-OV-mRNA data/Figures mRNA/htree_tcganal2.pdf", height=80, width=100) #išsaugijimui didesniu formatu
plot(dend5) #dstant samples should be excluded #save portrait 20x66 inch pdf
dev.off()
```

```{r}
#using norm.counts
#clinical feature: gtex or TCGA data
dim(norm.counts) #buvo tiek:494 4684
train_response <- as.factor(coldata_gtcga$gtex)
res_gtex = cv.glmnet(
  x = norm.counts,
  y = train_response,
  alpha = 1,
  family = "binomial"
)
res_gtex #atrenka 8, dabar 9
# Getting genes that contribute for the prediction
res_coef_gtex = coef(res_gtex, s="lambda.min") # the "coef" function returns a sparse matrix
head(res_coef_gtex) # in a sparse matrix the "." represents the value of zero
# get coefficients with non-zero values
res_coef_gtex = res_coef_gtex[res_coef_gtex[,1] != 0,] 
# note how performing this operation changed the type of the variable
# remove first coefficient as this is the intercept, a variable of the model itself
res_coef_gtex = res_coef_gtex[-1]
res_coef_gtex_names = names(res_coef_gtex) # get names of the (non-zero) variables.
length(res_coef_gtex_names) # number of selected genes 
```
```{r}
#convert relavant gene names
res_coef_gtex_names <- gsub("\\..*", "",res_coef_gtex_names)
res_coef_gtex_names
#biomart
gtcga_genes <- getBM(attributes = c('ensembl_gene_id','external_gene_name'), #values to retreve
      filters = "ensembl_gene_id", #input on quary
      values = res_coef_gtex_names,
      mart = ensembl.con) #sukurtas conection object
gtcga_genes
```
  ensembl_gene_id external_gene_name
1 ENSG00000170421               KRT8    keratinas
2 ENSG00000182196            ARL6IP4 - iš kitų 2 norma
3 ENSG00000184205             TSPYL2 - iš kitų 2 norm
4 ENSG00000187800              PEAR1    Platelet Endothelial Aggregation Receptor 
5 ENSG00000189143              CLDN4    klaudinas, tight junction 
6 ENSG00000196440             ARMCX4    Armadillo Repeat Containing X-Linked 4
7 ENSG00000265681              RPL17 -  iš gdc norm
8 ENSG00000272189           MAP7-AS1    microtubule geno lncRNA
9 ENSG00000283498          MIR1244-2    mikroRNR genas

palyginimas 
```{r}
deseq_vardai <- gtcga_genes$external_gene_name
intersect(deseq_vardai, gdc_vardai)
intersect(tcga_vardai, deseq_vardai)
```
# kuo skirias deseq kai duodu jau tą patį modelį kaip po to GLM?
```{r}
#is this nesąmonė?
# create dds
dds_gtex <- DESeqDataSetFromMatrix(countData = counts_gtcga,
                              colData = coldata_gtcga,
                              design = ~ gtex ) # not spcifying model

## remove all genes with counts < 15 in more than 75% of samples (488*0.75=366)
## suggested by WGCNA on RNAseq FAQ

dds75_dds_gtex <- dds_gtex[rowSums(counts(dds_gtex) >= 15) >= 366,] #čia nuo meginiu sk!
nrow(dds75_dds_gtex) # 4684 genes


# perform variance stabilization
dds_norm_gtex <- vst(dds75_dds_gtex) 


# get normalized counts
norm.counts_gtex <- assay(dds_norm_gtex) %>% 
  t()
```
lieka sveiki skaičiai dds_gtex, bet kai cauni normalized counts, much different skaičiai

```{r}
dds <- DESeqDataSetFromMatrix(countData = counts_gtcga,
                              colData = coldata_gtcga,
                              design= ~ gtex)
dds <- DESeq(dds)
resultsNames(dds) # lists the coefficients
res <- results(dds, name="gtex_ovarian.cancer_vs_control")
# or to shrink log fold changes association with condition:
resNorm <- lfcShrink(dds, coef=2, type="normal") #čia neveikia apeglm, todėl normal 
sum(res$padj < 0.1, na.rm=TRUE) #20990 tiek genų dge
res05 <- results(dds, alpha=0.05)
summary(res05)
plotMA(res, ylim=c(-2,2))


par(mfrow=c(1,3), mar=c(4,4,2,1))
xlim <- c(1,1e5); ylim <- c(-3,3)
plotMA(resNorm, xlim=xlim, ylim=ylim, main="normal")
```
```{r}
plotCounts(dds, gene=which.min(res$padj), intgroup="gtex") #išrinko geną madąl (mitotic arest gene)
```



